---
layout: paper
title: Cross-modality mapping using image varifolds to align tissue-scale atlases to molecular-scale measures with application to 2D brain sections
authors: Kaitlin M. Stouffer, Alain Trouv√©, Laurent Younes, Michael Kunst, Lydia Ng, Hongkui Zeng, <b><u>Manjari Anant</u></b>, <b>Jean Fan</b>, Yongsoo Kim, Xiaoyin Chen, Mara Rue, Michael I. Miller
abstract: This paper explicates a solution to building correspondences between molecular-scale transcriptomics and tissue-scale atlases. This problem arises in atlas construction and cross-specimen/technology alignment where specimens per emerging technology remain sparse and conventional image representations cannot efficiently model the high dimensions from subcellular detection of thousands of genes. We address these challenges by representing spatial transcriptomics data as generalized functions encoding position and high-dimensional feature (gene, cell type) identity. We map onto low-dimensional atlas ontologies by modeling regions as homogeneous random fields with unknown transcriptomic feature distribution. We solve simultaneously for the minimizing geodesic diffeomorphism of coordinates through LDDMM and for these latent feature densities. We map tissue-scale mouse brain atlases to gene-based and cell-based transcriptomics data from MERFISH and BARseq technologies and to histopathology and cross-species atlases to illustrate integration of diverse molecular and cellular datasets into a single coordinate system as a means of comparison and further atlas construction.
journal: Nature Communications. April 25, 2024. doi.org/10.1038/s41467-024-47883-4
journalurl: https://www.nature.com/articles/s41467-024-47883-4
pubmedurl: https://pubmed.ncbi.nlm.nih.gov/38664422/
---

