---
layout: paper
title: STalign - alignment of spatial transcriptomics data using diffeomorphic metric mapping
authors: <b><u>Kalen Clifton*</u></b>, <b><u>Manjari Anant*</u></b>, <b><u>Gohta Aihara</u></b>, <b><u>Lyla Atta</u></b>, Osagie K Aimiuwu, Justus M Kebschull, Michael I Miller, Daniel Tward^, <b>Jean Fan^</b>
image: /assets/papers/STalign_anim_small.gif
abstract: Spatial transcriptomics (ST) technologies enable high throughput gene expression characterization within thin tissue sections. However, comparing spatial observations across sections, samples, and technologies remains challenging. To address this challenge, we developed STalign to align ST datasets in a manner that accounts for partially matched tissue sections and other local non-linear distortions using diffeomorphic metric mapping. We apply STalign to align ST datasets within and across technologies as well as to align ST datasets to a 3D common coordinate framework. We show that STalign achieves high gene expression and cell-type correspondence across matched spatial locations that is significantly improved over manual and landmark-based affine alignments. Applying STalign to align ST datasets of the mouse brain to the 3D common coordinate framework from the Allen Brain Atlas, we highlight how STalign can enable the interrogation of compositional heterogeneity across anatomical structures. 
journal: Nature Communications. December 8, 2023. doi.org/10.1038/s41467-023-43915-7
journalurl: https://www.nature.com/articles/s41467-023-43915-7
pubmedurl: https://pubmed.ncbi.nlm.nih.gov/38065970/
code: https://github.com/JEFworks-Lab/STalign
---
