---
layout: paper
title: Opportunities and challenges of single-cell and spatially resolved genomics methods for neuroscience discovery
authors: Boyan Bonev, Gon√ßalo Castelo-Branco, Fei Chen, Simone Codeluppi, M. Ryan Corces, <b>Jean Fan</b>, Myriam Heiman, Kenneth Harris, Fumitaka Inoue, Manolis Kellis, Ariel Levine, Mo Lotfollahi, Chongyuan Luo, Kristen R. Maynard, Mor Nitzan, Vijay Ramani, Rahul Satijia, Lucas Schirmer, Yin Shen, Na Sun, Gilad S. Green, Fabian Theis, Xiao Wang, Joshua D. Welch, Ozgun Gokce, Genevieve Konopka, Shane Liddelow, Evan Macosko, Omer Ali Bayraktar, Naomi Habib, Tomasz J. Nowakowski
abstract: Over the past decade, single-cell genomics technologies have allowed scalable profiling of cell-type-specific features, which has substantially increased our ability to study cellular diversity and transcriptional programs in heterogeneous tissues. Yet our understanding of mechanisms of gene regulation or the rules that govern interactions between cell types is still limited. The advent of new computational pipelines and technologies, such as single-cell epigenomics and spatially resolved transcriptomics, has created opportunities to explore two new axes of biological variation - cell-intrinsic regulation of cell states and expression programs and interactions between cells. Here, we summarize the most promising and robust technologies in these areas, discuss their strengths and limitations and discuss key computational approaches for analysis of these complex datasets. We highlight how data sharing and integration, documentation, visualization and benchmarking of results contribute to transparency, reproducibility, collaboration and democratization in neuroscience, and discuss needs and opportunities for future technology development and analysis.
journal: Nature Neuroscience. December 3, 2024. doi.org/10.1038/s41593-024-01806-0
journalurl: https://www.nature.com/articles/s41593-024-01806-0
pubmedurl: https://pubmed.ncbi.nlm.nih.gov/39627587/
---